src/Tools/Pagination/Paginator.php

```diff
@@ -36,13 +47,26 @@ class Paginator implements Countable, IteratorAggregate
     public const HINT_ENABLE_DISTINCT = 'paginator.distinct.enable';
 
     private readonly Query $query;
-    private bool|null $useOutputWalkers = null;
-    private int|null $count             = null;
+    private bool|null $useResultQueryOutputWalker = null;
+    private bool|null $useCountQueryOutputWalker  = null;
+    private int|null $count                       = null;
+    /**
+     * The auto-detection of queries style was added a lot later to this class, and this
+     *  class historically was by default using the more complex queries style, which means that
+     *  the simple queries style is potentially very under-tested in production systems. The purpose
+     *  of this variable is to not introduce breaking changes until an impression is developed that
+     *  the simple queries style has been battle-tested enough.
+     */
+    private bool $queryStyleAutoDetectionEnabled = false;
+    private bool $queryCouldHaveToManyJoins      = true;
 
-    /** @param bool $fetchJoinCollection Whether the query joins a collection (true by default). */
+    /**
+     * @param bool $queryCouldProduceDuplicates Whether the query could produce partially duplicated records. One case
+     *  when it does is when it joins a collection.
+     */
     public function __construct(
         Query|QueryBuilder $query,
-        private readonly bool $fetchJoinCollection = true,
+        private readonly bool $queryCouldProduceDuplicates = true,
```

Comment:
The docs need to be updated at the very least because of this. Can you also read it and see if anything else needs to be modified, and whether any new behavior you are adding is worth documenting?